walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,136 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.List;
+import java.util.ArrayList;
+import com.google.sps.Event;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
+
+  /** Return a collection of time periods where all attendees in `request` are free
+   * @param events - a collection of events to be scheduled for that day 
+   * @param request - the details of a meeting proposal, including the attendees of 
+                      that event
+   * @return a collection of times where all attendees are free
+   **/
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    // Print out input for debugging
+    for (Event event: events) {
+        System.out.println(event);
+    }
+    System.out.println("DEBUG \n");
+
+    // Add events scheduled with mandatory and optional guests
+    // Assuming mandatory guests can't be scheduled for 2 events
+    Collection<TimeRange> timeRanges = new ArrayList<TimeRange>();
+    for (Event event: events) {
+        if (request.getAttendees().containsAll(event.getAttendees())) { // check event attendees are a subset of request attendees
+            timeRanges.add(event.getWhen());
+        }
+    }
+
+    // special case - event schedule at beginning for day
+    TimeRange eventAtStartOfDay = eventAtStartOfDay(timeRanges);
+    int startTime = (eventAtStartOfDay != null) ? eventAtStartOfDay.end() : TimeRange.START_OF_DAY;
+    
+    Collection<TimeRange> freePeriods = new ArrayList<TimeRange>();
+    while (startTime < TimeRange.END_OF_DAY) {
+        int endTime = getEndTime(startTime, timeRanges); // find upper limit of free time
+        // add to free times
+        TimeRange proposedTime = TimeRange.fromStartEnd(startTime, endTime, true);
+        freePeriods.add(proposedTime);
+        System.out.println("Proposed time: " + proposedTime);
+        
+        if (endTime >= TimeRange.END_OF_DAY)
+            break;
```

Comment:
This structure is quite wired. You could make it either 
if (endTime >= TimeRange.END_OF_DAY) {
  break;
}
or
if (endTime >= TimeRange.END_OF_DAY) break;